### ISSUE TITLE
optional setters

### ISSUE THREAD

**Issue by colin-kiegel** [colin-kiegel/rust-derive-builder#36]
Right now all setters are mandatory. I.e. the build method of `derive_builder` will error if not all fields are initialized.

# use case

User wants to make some setters optional. If no value is provided `Default::default()` should be used instead.

Outlook: [custom defaults](https://github.com/colin-kiegel/rust-derive-builder/issues/3)

# possible api

field annotation:
- optional: `#[builder(optional="true")]` or `#[builder(mandatory="false")]` 
- mandatory `#[builder(optional="false")]` or `#[builder(mandatory="true")]`

shorthand
- optional: `#[builder(optional)]`
- mandatory: `#[builder(mandatory)]`

# example

``` rust
#[derive(Builder)]
struct Foo {
    #[builder(optional)]
    foo: bool
}

#[test]
fn test() {
    let x = FooBuilder::default().build().unwrap();
                                       // ^~~ currently panics with `foo uninitialized`
    assert_eq!(x.foo, false);
}
```

# open questions

- API (see discussion)
- What should be the default behaviour? (See [this suggestion](https://github.com/colin-kiegel/rust-derive-builder/issues/36#issuecomment-281347160) to switch back to "optional by default").

# discussion

Conceptually the hardest part is to pick a good api (=annotation). :-)

This also has significant overlap with the idea to have [custom defaults](https://github.com/colin-kiegel/rust-derive-builder/issues/3), e.g. I would expect that `custom defaults` implicitly make a setter optional. `custom defaults` wouldn't even make sense for mandatory setters. You could argue that an explicit notation for `optional fields` is not needed.

However most types implement `Default`, so a custom default may be unnecessary. Additionally the `optional field` annotation could be used for the whole struct, instead of individual fields - thereby inverting the opt-in to opt-out for that struct. In these cases an optional field annotation is more convenient than custom defaults - but it doesn't remove the overlap.

Now comes the hard part.

Should it be ...
- `optional="true"` vs. `optional="false"`
- .. the above with mandatory instead of optional
- `optional` vs `mandatory`?
- or the whole lot: `optional`, `optional="true"`, `mandatory="false"` vs `mandatory`, `mandatory="true"`, `optional="false"`
- or something more _extensible_ like `style=optional` vs. `style=mandatory`  (where `style` is only a placeholder name)?

Right now I can't think of a reason to extend the binary set `optional` and `mandatory`. 

My tendency is to allow the whole lot of `optional`, `optional="true"`, `mandatory="false"` vs _inverse_. The main drawback I see is that it may clog the documentation a bit - at least if we want to cover all ways to do something.

# alternatives

- mandatory (_inverse_)
- mandatory + optional
- use_default
- default (<- this may collides with the api for [custom defaults](https://github.com/colin-kiegel/rust-derive-builder/issues/3))
- style=... (more open)
- voluntary instead of optional
- ...

**Comment by colin-kiegel** [comment]
We could also put it in a `setter` namespace like ` #[builder(setter(voluntary))]`. I am not that reluctant to use nested namespaces anymore, if it helps us in making the options more understandable and future proof.
* Do we say something about the _setter_ or the _field_ or the _relation of the builder to that field_?

I feel like it's more about the _relation of the builder to that field_, since the error is triggered in the `build` method. If the user implements an alternative build method, he will not see any effect of `voluntary` vs. `mandatory`. This means that _if_ we want to namespace this setting it should probably go into a namespace for the build method.

Opinions about all of this are very welcomed. :-)

**Comment by killercup** [comment]
The opposite of _mandatory_ is _optional_ IMO. Though voluntary has a nice ring to it 😄

**Comment by JuanPotato** [comment]
This would be very useful for struct values that are `Option<...>` which is where I found myself wanting to use this crate but then found that I'd be doing `.thing(None)` for anything that I didn't want.

**Comment by colin-kiegel** [comment]
@killercup Updated the proposal from `voluntary` -> `optional`. Thanks for the hint. :-)

**Comment by shssoichiro** [comment]
Glad to see this on the radar! I was using 0.2.x of this crate, which had this as the default behavior, to build a 30-field struct where most of the fields are optional, and was surprised to see this was removed in 0.3.0, which makes version 0.3 unusable to me. I personally think the default behavior should be to default any unspecified fields to their `Default` value, and users can opt-in to making all fields required. It seems like a more common use case to me, although I may be biased because of my personal use case.

**Comment by colin-kiegel** [comment]
@shssoichiro thx for your feedback - I'll add that to open questions. :-)

**Comment by colin-kiegel** [comment]
@TedDriggs commented in https://github.com/colin-kiegel/rust-derive-builder/issues/3#issuecomment-284820123 that `#[builder(default)]` would be a nice shorthand to use `Default::default()`. However `#[builder(default)]` should mark these fields implicitly optional and take the role of `#[builder(optional)]` of this proposal - at least on the field-level.

But how would that relate to the other aspects of this proposal?

I feel like there are different perspectives to it.
- mixing `#[builder(default)]` and `#[builder(default = "foo()")]` it feels very natural.
- mixing `#[builder(default)]` and `#[builder(mandatory)]` feels weird - they don't feel like natural opposites of each other.

As far as I can see there might be 3 to 4 scenarios
- all fields optional
- some fields optional, others mandatory (for ergonomic reasons this could be splitted into two sub-cases: (a) mostly optional vs. (b) mostly mandatory fields)
- all fields mandatory

The _pure_ cases (all fields optional / mandatory) would need some annotation on the _struct_-level. Whereas the _inbetween_ cases would need additional annotation on the _field_-level and _possibly_ also annotation on the struct-level to define the _starting point_?

Up to this point this looks pretty symmetric. However, when we start to consider the return type of the build function, the case 'all fields optional' is special. It's the only case which can return the target struct type `Foo` immediately without error handling. All other cases need to return `Result<Foo, ...>` or panic.

The question is: Should the return type of the build method be implicitly controlled? Or in other words, would we allow this?
```rust
#[derive(Builder)]
#[builder(optional)] // or #[builder(default)] <-- all fields optional by default -
                     // looks like the return value is `Foo`
struct Foo {
    foo1: /* ... */,
    foo2: /* ... */,
    #[builder(mandatory)] // <-- this field is not optional -
                          // oh no, now the return type is `Result<Foo, ...>` :-(
    foo3: /* ... */,
}
```

Another more subtle example
```rust
#[derive(Builder)] // <-- all fields mandatory by default - 
                   // looks like the return type is `Result<Foo, ...>`
struct Foo {
    #[builder(default)]
    foo1: /* ... */,
    #[builder(default)]
    foo2: /* ... */,
    // #[builder(default)] // <-- but what if we uncomment this line?
                           // All fields would be optional. Would that implicitly
                           // switch the return type to `Foo`, because we don't
                           // need error handling anymore? :-/
    foo3: /* ... */,
}
```

When I think about it this way, I feel like we should have a top-level attribute which controls the _signature_ of the `build` method. All field-specific attributes should then respect this signature. I think this is already an important constraint. It would mean that we would probably also name the according attributes on the struct-level different from those on the field level.

Also I think we have very strong arguments to use `#[builder(default)]` on the field level. Just with that we already can express the purely mandatory case and the mixed case.

The only thing missing is the purely optional case.
```rust
#[derive(Builder)]
#[builder(optional)] // <-- all fields will be optional
                     // no way to opt-out for individual fields
                     // return value is definitely `Foo`
struct Foo {
    foo1: /* ... */, // `#[builder(default)]` is implied and can be omitted (?)
    foo2: /* ... */,
    #[builder(default = "foo()")] // but we can pick a different default value
    foo3: /* ... */,
}
```

I.e. the attribute `optional` operates on the struct and sets the `build()` return type. The attribute `default [= "foo()"]` operates on fields only and is implied for the `optional` case.

Drawback: The mixed case with mostly optional fields would be a bit verbose.

Alternatives: Maybe still another name for `optional`.

How does that sound?

**Comment by TedDriggs** [comment]
@colin-kiegel maybe I'm missing something, but wouldn't the #[builder(default)] attribute in isolation be sufficient? If the struct has that attribute, then any field which isn't explicitly initialized will get the value from the struct's default instance. If a field has the attribute, then it can be omitted to get the default value.

**Comment by colin-kiegel** [comment]
@TedDriggs Yes, `#[builder(default)]` as in the current PR #53 would be sufficient to express (almost) everything that is mentioned here. It has not as many different ways to express things, but it's still fully expressive - that's probably a good thing to start with. The only thing it can't express is, whether the `build()` method should return `Result<Foo, String>` or `Foo`. My current thinking is to handle that as a separate feature.

I.e. with the current PR, the `build()` method would always have the signature `Result<Foo, String>` even if all fields are optional and it _cannot_ fail. The reason behind this is, that future extensions like custom validations per field could introduce other failure modes.

**Comment by TedDriggs** [comment]
> I.e. with the current PR, the `build()` method would always have the signature `Result<Foo, String>` even if all fields are optional and it cannot fail. The reason behind this is, that future extensions like custom validations per field could introduce other failure modes.

That sounds perfect; it also would allow for the existence of a `Builder<T>` trait which worked for all builders, even if every field was optional.

If there's a strong need for the "build-or-panic" case, then that could be `build` and the Result-producing one could be `try_build`, I suppose. Or the 'build-or-panic' method could be `build_unwrap()` to make clearer what it was doing.

**Comment by colin-kiegel** [comment]
@TedDriggs Thx for your feedback! That helps me to validate that I am on the right track.

PS: I am waiting for [`TryFrom<T>` and `TryInto<T>`](https://github.com/rust-lang/rfcs/blob/master/text/1542-try-from.md) to become stable. These would be the idiomatic builder traits IMO. I just created a ticket _whether we should change the current `build` name to `try_into`_, you can join the discussion there: https://github.com/colin-kiegel/rust-derive-builder/issues/54. :-)

**Comment by colin-kiegel** [comment]
This is implemented via #3.

The question of the `build()` method signature is discussed in #56.
